src/awkward/_connect/pyarrow_table_conv.py

```diff
@@ -0,0 +1,250 @@
+        new_field = pyarrow.field(
+            aafield.name, type=typ.storage_type, nullable=aafield.nullable
+        )
+        return new_field
+
+    # We have nested types
+    native_fields = [
+        awkward_arrow_field_to_native(field)  # Recurse
+        for field in fields
+    ]
+    native_type = _make_pyarrow_type_like(typ, native_fields)
+    new_field = pyarrow.field(aafield.name, type=native_type, nullable=aafield.nullable)
+    return new_field
+
+
+def native_arrow_field_to_akarraytype(
+    ntv_field: pyarrow.Field, metadata: dict
+) -> pyarrow.Field:
+    if isinstance(ntv_field, AwkwardArrowType):
+        raise ValueError(f"field {ntv_field} is already an AwkwardArrowType")
+    storage_type = ntv_field.type
+    fields = _fields_of_strg_type(storage_type)
+    if len(fields) > 0:
+        # We need to replace storage_type with one that contains AwkwardArrowTypes.
+        awkwardized_fields = [
+            native_arrow_field_to_akarraytype(field, meta)  # Recurse
+            for field, meta in zip(fields, metadata["subfield_metadata"])
+        ]
+        storage_type = _make_pyarrow_type_like(storage_type, awkwardized_fields)
+    ak_type = AwkwardArrowType._from_metadata_object(storage_type, metadata)
+    return pyarrow.field(ntv_field.name, type=ak_type, nullable=ntv_field.nullable)
+
+
+def _fields_of_strg_type(typ: pyarrow.Type) -> list[pyarrow.Field]:
+    if isinstance(typ, pyarrow.lib.DictionaryType):
+        return [
+            pyarrow.field("value", typ.value_type)
+        ]  # Wrap in a field for consistency
+    return [typ.field(i) for i in range(typ.num_fields)]
+
+
+def _make_pyarrow_type_like(
+    typ: pyarrow.Type, fields: list[pyarrow.Field]
+) -> pyarrow.Type:
+    storage_type = to_awkwardarrow_storage_types(typ)[1]
+    if isinstance(storage_type, pyarrow.lib.DictionaryType):
+        return pyarrow.dictionary(storage_type.index_type, fields[0].type)
+    if isinstance(storage_type, pyarrow.lib.FixedSizeListType):
+        return pyarrow.list_(fields[0], storage_type.list_size)
+    if isinstance(storage_type, pyarrow.lib.ListType):
+        return pyarrow.list_(fields[0])
+    if isinstance(storage_type, pyarrow.lib.LargeListType):
+        return pyarrow.large_list(fields[0])
+    if isinstance(storage_type, pyarrow.lib.MapType):
+        # return pyarrow.map_(storage_type.index_type, fields[0])
+        raise NotImplementedError("pyarrow MapType is not supported by Awkward")
+    if isinstance(storage_type, pyarrow.lib.StructType):
+        return pyarrow.struct(fields)
+    if isinstance(storage_type, pyarrow.lib.UnionType):
+        return pyarrow.union(fields, storage_type.mode, storage_type.type_codes)
+    if isinstance(storage_type, pyarrow.lib.DataType) and storage_type.num_fields == 0:
+        # Catch-all for primitive types, nulltype, string types, FixedSizeBinaryType
+        return storage_type
+    raise NotImplementedError(f"Type {typ} is not handled for conversion.")
```

Comment:
I agree with you here. I think I wrote it this way out of habit or maybe fear that ruff would change it to this anyway. But it was fine with the if-elif chain. I left the final `raise NotImplementedError` out of the chain because that's more bulletproof. If you ever make it out of that that chain without returning something, fail.